### PR TITLE
601 complete support for replication across threads

### PR DESCRIPTION
**Pull request by vdimiczrc**

### CODE REVIEW

File: include/alp/omp/matrix.hpp [hunk 1]
```diff
@@ -59,11 +59,9 @@ namespace alp {
 		>::template change_backend< config::default_sequential_backend >::type
 		get_view( SourceMatrix &source, const size_t tr, const size_t tc, const size_t rt, const size_t br, const size_t bc ) {
 
-			(void) rt;
-
 			// get the container
 			const auto &distribution = getAmf( source ).getDistribution();
-			const size_t thread_id = tr * distribution.getThreadGridDims().second + tc;
+			const size_t thread_id = distribution.getThreadId( tr, tc, rt );
```
Comment (hidanielesgit) [inline include/alp/omp/matrix.hpp:64]: based on thread's coord

File: include/alp/omp/matrix.hpp [hunk 2]
```diff
@@ -59,11 +59,9 @@ namespace alp {
...
 			const size_t block_id = br * distribution.getLocalBlockGridDims( tr, tc ).second + bc;
```
Comment (hidanielesgit) [inline include/alp/omp/matrix.hpp:65]: same here

File: include/alp/omp/matrix.hpp [hunk 3]
```diff
@@ -59,11 +59,9 @@ namespace alp {
 		>::template change_backend< config::default_sequential_backend >::type
 		get_view( SourceMatrix &source, const size_t tr, const size_t tc, const size_t rt, const size_t br, const size_t bc ) {
```
Comment (hidanielesgit) [inline include/alp/omp/matrix.hpp:60]: single thread coords object

File: include/alp/omp/storage.hpp [hunk 4]
```diff
@@ -203,19 +218,24 @@ namespace alp {
...
-			std::pair< size_t, size_t > getThreadGridDims() const {
-				return { Tr, Tc };
+			const ThreadGrid getThreadGridDims() const {
+				return ThreadGrid( Tr, Tc );
 			}
 
 			/** Returns the thread ID corresponding to the given thread coordinates. */
-			size_t getThreadId( const size_t tr, const size_t tc ) const {
-				return tr * Tc + tc;
+			size_t getThreadId( const size_t tr, const size_t tc, const size_t rt ) const {
```
Comment (hidanielesgit) [inline include/alp/omp/storage.hpp:237]: single thread coord object

File: include/alp/omp/storage.hpp [hunk 5]
```diff
@@ -225,7 +245,7 @@ namespace alp {
 
 			/** Returns the dimensions of the block grid associated to the given thread */
 			std::pair< size_t, size_t > getLocalBlockGridDims( const size_t tr, const size_t tc ) const {
```
Comment (hidanielesgit) [inline include/alp/omp/storage.hpp:247]: same here

File: include/alp/omp/storage.hpp [hunk 6]
```diff
@@ -403,7 +424,7 @@ namespace alp {
 					const typename Distribution::GlobalCoord global( imf_r.map( i ), imf_c.map( j ) );
 					const typename Distribution::LocalCoord local = distribution.mapGlobalToLocal( global );
 
-					const size_t thread = local.tr * distribution.getThreadGridDims().second + local.tc;
+					const size_t thread = local.tr * distribution.getThreadGridDims().Tc + local.tc;
```
Comment (hidanielesgit) [inline include/alp/omp/storage.hpp:427]: within (25D_?)Distribution consider: 
- having ThreadGrid object constructed 
- exchanging ThreadCoords for specific threads in the grid

File: include/alp/omp/vector.hpp [hunk 7]
```diff
@@ -151,7 +151,7 @@ namespace alp {
 				Vector(
 					const Distribution &d,
 					const size_t cap = 0
-				) : num_buffers( d.getThreadGridDims().first * d.getThreadGridDims().second ),
+				) : num_buffers( d.getThreadGridDims().Tr * d.getThreadGridDims().Tc * d.getThreadGridDims().Rt ),
```
Comment (hidanielesgit) [inline include/alp/omp/vector.hpp:154]: The distribution could provide the total number of threads in the grid?

File: include/alp/omp/vector.hpp [hunk 8]
```diff
@@ -169,9 +169,8 @@ namespace alp {
 
 					#pragma omp parallel for
 					for( size_t thread = 0; thread < config::OMP::current_threads(); ++thread ) {
-						const size_t tr = d.getThreadCoords( thread ).first;
-						const size_t tc = d.getThreadCoords( thread ).second;
-						const auto block_grid_dims = d.getLocalBlockGridDims( tr, tc );
+						const auto t_coords = d.getThreadCoords( thread );
+						const auto block_grid_dims = d.getLocalBlockGridDims( t_coords.tr, t_coords.tc );
```
Comment (hidanielesgit) [inline include/alp/omp/vector.hpp:173]: passing thread coord object

File: include/alp/omp/vector.hpp [hunk 9]
```diff
@@ -202,7 +201,7 @@ namespace alp {
 						// Populate the array of internal container wrappers
 						for( size_t br = 0; br < block_grid_dims.first; ++br ) {
 							for( size_t bc = 0; bc < block_grid_dims.second; ++bc ) {
-								const size_t offset = d.getBlocksOffset( tr, tc, br, bc );
+								const size_t offset = d.getBlocksOffset( t_coords.tr, t_coords.tc, br, bc );
```
Comment (hidanielesgit) [inline include/alp/omp/vector.hpp:204]: passing thread coord object

**Review by hidanielesgit — CHANGES_REQUESTED** [review]

File: include/alp/omp/matrix.hpp [hunk 10]
```diff
@@ -50,21 +50,20 @@ namespace alp {
...
-
-			(void) rt;
+		get_view( SourceMatrix &source, const ThreadCoords t, const size_t br, const size_t bc ) {
 
 			// get the container
 			const auto &distribution = getAmf( source ).getDistribution();
-			const size_t thread_id = tr * distribution.getThreadGridDims().second + tc;
-			const size_t block_id = br * distribution.getLocalBlockGridDims( tr, tc ).second + bc;
+			const size_t thread_id = distribution.getThreadId( t );
+			const size_t block_id = br * distribution.getLocalBlockGridDims( t ).second + bc;
```
Comment (hidanielesgit) [inline include/alp/omp/matrix.hpp:66]: Leave a comment saying that the distribution could compute the block_id?

File: include/alp/omp/storage.hpp [hunk 11]
```diff
@@ -403,9 +417,8 @@ namespace alp {
 					const typename Distribution::GlobalCoord global( imf_r.map( i ), imf_c.map( j ) );
 					const typename Distribution::LocalCoord local = distribution.mapGlobalToLocal( global );
 
-					const size_t thread = local.tr * distribution.getThreadGridDims().second + local.tc;
-
-					const size_t local_block = local.br * distribution.getLocalBlockGridDims( local.tr, local.tc ).second + local.bc;
+					const size_t thread = distribution.getThreadId( local.getThreadCoords() );
+					const size_t local_block = local.br * distribution.getLocalBlockGridDims( local.getThreadCoords() ).second + local.bc;
```
Comment (hidanielesgit) [inline include/alp/omp/storage.hpp:421]: If logic for computing block_id goes into the distribution than this could also be adapted accordingly

**Review by hidanielesgit — APPROVED** [review]

**Review by djelovina — APPROVED** [review]

**Pull request merged**